core-rust/state-manager/src/staging/cache.rs

```diff
@@ -312,6 +312,57 @@ impl<'s, S: SubstateDatabase> SubstateDatabase for StagedStore<'s, S> {
         }
     }
 
+    fn list_entries_from(
+        &self,
+        partition_key: &DbPartitionKey,
+        from_sort_key: Option<&DbSortKey>,
+    ) -> Box<dyn Iterator<Item = PartitionEntry> + '_> {
+        let partition_updates = self.overlay.partition_updates.get(partition_key);
+        let Some(partition_updates) = partition_updates else {
+            return self.root.list_entries_from(partition_key, from_sort_key);
+        };
+
+        let cloned_from_sort_key = from_sort_key.cloned();
+
+        match partition_updates {
+            ImmutablePartitionUpdates::Delta { substate_updates } => {
+                let overlaid_iter = substate_updates
+                    .iter()
+                    .map(|(sort_key, update)| (sort_key.clone(), update.clone()))
+                    .sorted_by(|(left_key, _), (right_key, _)| left_key.cmp(right_key))
+                    .skip_while(move |(key, _)| {
+                        if let Some(from_sort_key) = &cloned_from_sort_key {
+                            key.lt(from_sort_key)
+                        } else {
+                            false
+                        }
+                    });
+
+                Box::new(SubstateOverlayIterator::new(
+                    self.root.list_entries_from(partition_key, from_sort_key),
+                    Box::new(overlaid_iter),
+                ))
+            }
+            ImmutablePartitionUpdates::Batch(batch) => match batch {
+                BatchImmutablePartitionUpdates::Reset {
+                    new_substate_values,
+                } => Box::new(
+                    new_substate_values
+                        .iter()
+                        .map(|(sort_key, update)| (sort_key.clone(), update.clone()))
+                        .sorted_by(|(left_key, _), (right_key, _)| left_key.cmp(right_key))
+                        .skip_while(move |(key, _)| {
+                            if let Some(from_sort_key) = &cloned_from_sort_key {
```

Comment:
(optional) if you dislike `if`s as much as I do, you can trade it for the `let from_sort_key = from_sort_key.cloned().unwrap_or(DbSortKey(vec![]));` trick that you applied at https://github.com/radixdlt/babylon-node/pull/746/files#diff-5cf22ad2b534dd277f51718bd8cf5c3e624cecbc4819906008c39e0798427ce3R1192